File: src/algorithms/linked-lists/insertion-deletion/delete-by-value/__tests__/delete-by-value_test.go
package main

import (
    "reflect"
    "testing"
)

func buildListDeleteByValue(values []int) *ListNode {
    var head *ListNode
    for idx := len(values) - 1; idx >= 0; idx-- {
        head = &ListNode{value: values[idx], next: head}
    }
    return head
}

func listToSliceDeleteByValue(head *ListNode) []int {
    result := []int{}
    for head != nil {
        result = append(result, head.value)
        head = head.next
    }
    return result
}

func TestDeleteByValueMiddle(t *testing.T) {
    result := deleteByValue(buildListDeleteByValue([]int{1, 2, 3, 4, 5}), 3)
    if !reflect.DeepEqual(listToSliceDeleteByValue(result), []int{1, 2, 4, 5}) {
        t.Error("expected [1 2 4 5] after deleting 3 from middle")
    }
}

func TestDeleteByValueHead(t *testing.T) {
    result := deleteByValue(buildListDeleteByValue([]int{1, 2, 3}), 1)
    if !reflect.DeepEqual(listToSliceDeleteByValue(result), []int{2, 3}) {
        t.Error("expected [2 3] after deleting head")
    }
}

func TestDeleteByValueLast(t *testing.T) {
    result := deleteByValue(buildListDeleteByValue([]int{1, 2, 3, 4}), 4)
    if !reflect.DeepEqual(listToSliceDeleteByValue(result), []int{1, 2, 3}) {
        t.Error("expected [1 2 3] after deleting last node")
    }
}

func TestDeleteByValueEmptyList(t *testing.T) {
    result := deleteByValue(nil, 5)
    if result != nil {
        t.Error("expected nil for empty list")
    }
}

func TestDeleteByValueTargetNotFound(t *testing.T) {
    result := deleteByValue(buildListDeleteByValue([]int{1, 2, 3}), 99)
    if !reflect.DeepEqual(listToSliceDeleteByValue(result), []int{1, 2, 3}) {
        t.Error("expected list unchanged when target not found")
    }
}

func TestDeleteByValueSingleNodeMatch(t *testing.T) {
    result := deleteByValue(buildListDeleteByValue([]int{7}), 7)
    if !reflect.DeepEqual(listToSliceDeleteByValue(result), []int{}) {
        t.Error("expected empty list after deleting only node")
    }
}

func TestDeleteByValueSingleNodeNoMatch(t *testing.T) {
    result := deleteByValue(buildListDeleteByValue([]int{7}), 5)
    if !reflect.DeepEqual(listToSliceDeleteByValue(result), []int{7}) {
        t.Error("expected [7] when single-node value differs from target")
    }
}

func TestDeleteByValueFirstOccurrenceOnly(t *testing.T) {
    result := deleteByValue(buildListDeleteByValue([]int{1, 2, 2, 3}), 2)
    if !reflect.DeepEqual(listToSliceDeleteByValue(result), []int{1, 2, 3}) {
        t.Error("expected only first occurrence deleted [1 2 3]")
    }
}

File: src/algorithms/linked-lists/insertion-deletion/delete-by-value/DeleteByValue_test.cpp
#include <cassert>
#include <vector>
#include "sources/DeleteByValue.cpp"

ListNode* buildList(const std::vector<int>& values) {
    ListNode* head = nullptr;
    for (int idx = static_cast<int>(values.size()) - 1; idx >= 0; idx--) {
        ListNode* node = new ListNode(values[idx]);
        node->next = head;
        head = node;
    }
    return head;
}

std::vector<int> listToVec(ListNode* head) {
    std::vector<int> result;
    while (head != nullptr) {
        result.push_back(head->value);
        head = head->next;
    }
    return result;
}

int main() {
    // deletes a node in the middle of the list
    assert(listToVec(deleteByValue(buildList({1, 2, 3, 4, 5}), 3)) == std::vector<int>({1, 2, 4, 5}));

    // deletes the head of the list
    assert(listToVec(deleteByValue(buildList({1, 2, 3}), 1)) == std::vector<int>({2, 3}));

    // deletes the last node in the list
    assert(listToVec(deleteByValue(buildList({1, 2, 3, 4}), 4)) == std::vector<int>({1, 2, 3}));

    // returns null for an empty list
    assert(deleteByValue(nullptr, 5) == nullptr);

    // returns the list unchanged when target is not found
    assert(listToVec(deleteByValue(buildList({1, 2, 3}), 99)) == std::vector<int>({1, 2, 3}));

    // deletes from a single-node list
    assert(listToVec(deleteByValue(buildList({7}), 7)) == std::vector<int>({}));

    // does not delete when single-node list value differs from target
    assert(listToVec(deleteByValue(buildList({7}), 5)) == std::vector<int>({7}));

    // deletes only the first occurrence
    assert(listToVec(deleteByValue(buildList({1, 2, 2, 3}), 2)) == std::vector<int>({1, 2, 3}));

    return 0;
}
